### ISSUE TITLE
Unexpected update caused by COLLATE

### ISSUE THREAD

**Issue by where2me** [pingcap/tidb#32382]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
DROP TABLE IF EXISTS t0;
CREATE TABLE IF NOT EXISTS t0(c0 NUMERIC);
INSERT INTO t0 VALUES(NULL);
UPDATE t0 SET c0 = 1 WHERE t0.c0 OR ('a' COLLATE 'latin1_bin');
```

### 2. What did you expect to see? (Required)
update should fail, as the where clause is evaluated to the value NULL.
### 3. What did you see instead (Required)
```sql
mysql> UPDATE t0 SET c0 = 1 WHERE t0.c0 OR ('a' COLLATE 'latin1_bin');
Query OK, 1 row affected (0.01 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```
### 4. What is your TiDB version? (Required)
mysql> SELECT tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()
                               |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.4.0
Edition: Community
Git Commit Hash: 55f3b24c1c9f506bd652ef1d162283541e428872
Git Branch: heads/refs/tags/v5.4.0
UTC Build Time: 2022-01-25 08:39:26
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

**Comment by solotzg** [comment]
Fixed in master branch
```
MySQL [test]> DROP TABLE IF EXISTS t0;
Query OK, 0 rows affected, 1 warning (0.00 sec)

MySQL [test]> CREATE TABLE IF NOT EXISTS t0(c0 NUMERIC);
Query OK, 0 rows affected (0.07 sec)

MySQL [test]> INSERT INTO t0 VALUES(NULL);
Query OK, 1 row affected (0.01 sec)

MySQL [test]> UPDATE t0 SET c0 = 1 WHERE t0.c0 OR ('a' COLLATE 'latin1_bin');
ERROR 1253 (42000): COLLATION 'latin1_bin' is not valid for CHARACTER SET 'utf8'
MySQL [test]> select version();
+----------------------------+
| version()                  |
+----------------------------+
| 5.7.25-TiDB-v6.1.0-nightly |
+----------------------------+
1 row in set (0.00 sec)
```

**Comment by zanmato1984** [comment]
This is a collation related issue. Changing to sig/sql-infra.

**Comment by bb7133** [comment]
@where2me For collations, we guarantee the compatibility only when `new_collations_enabled_on_first_bootstrap = true`. 

@solotzg I'm not sure why, but I got the same error message as MySQL:

```
tidb> UPDATE t0 SET c0 = 1 WHERE t0.c0 OR ('a' COLLATE 'utf8_bin');
ERROR 1292 (22007): Truncated incorrect DOUBLE value: 'a'

tidb> UPDATE t0 SET c0 = 1 WHERE t0.c0 OR (_latin1'a' COLLATE 'latin1_bin');
ERROR 1292 (22007): Truncated incorrect DOUBLE value: 'a'
tidb>
tidb> desc t0;
+-------+---------------+------+------+---------+-------+
| Field | Type          | Null | Key  | Default | Extra |
+-------+---------------+------+------+---------+-------+
| c0    | decimal(10,0) | YES  |      | NULL    |       |
+-------+---------------+------+------+---------+-------+
1 row in set (0.00 sec)
```

**Comment by where2me** [comment]
I found this case in TiDB v5.4.0.
In the latest release version, https://github.com/pingcap/tidb/releases/tag/v5.4.2
I get the expected result as follows.
```cmd 
mysql> UPDATE t0 SET c0 = 1 WHERE t0.c0 OR ('a' COLLATE 'latin1_bin');
ERROR 1292 (22007): Truncated incorrect DOUBLE value: 'a'
```
It seems to be fixed, as in TiDB v5.4.0 UPDATE succeeded.
